### DISCUSSION TITLE
ISO 7816 - add Calypso dump for proxmark

### DISCUSSION THREAD

**Discussion by hackcrypto** [metrodroid/metrodroid#806] [Q&A]
Hello devs,

I have a question about ISO 7816, how to dump or view full raw content of card for add to proxmark project ?

Thanks in advance !

**Comment by micolous** [comment]
If all you want is to dump the card on your PC:

* **easy option**: buy a cheap PC/SC compatible reader (eg: ACR122U)
* **hard option**: implement a PC/SC compatible reader API for Proxmark, like every nearly other Smartcard reader

That way you can just run the Java version of Metrodroid, and use its dump code.

Using a Proxmark to dump these cards is **totally** unnecessary:

* the cards are normally freely readable, without any keys
* there is very little published research into security issues with this card media (compared to MIFARE Classic)
* reconstructing an ISO 7816 card dump from a sniffer is tedious

Calypso is not some simple flash memory chip like MIFARE Classic – it has a filesystem API (ISO 7816-4) and a bunch of extensions for executing transactions.

If you want to really **write** this yourself, that's also **hard**, but to start:

1. Get a copy of ISO 7816-4. It's needed to understand all the terminology used for card communication.
2. Metrodroid's Calypso dump code can be found here: https://github.com/metrodroid/metrodroid/blob/3aa6dd31cfd2740bae85f07d9c8bd602e3863e1e/src/commonMain/kotlin/au/id/micolous/metrodroid/card/calypso/CalypsoApplication.kt#L206-L244

On a high level, what it's doing is:

1. Select the Calypso application (this step happens before this function)
2. Dump all the SFIs (special kind of small file)
3. For each possible Calypso file: select that file and then (try to) read its content

There is quite a lot of abstraction in that code – each line corresponds to multiple APDUs, and automatically stores the response in a mutable "capsule". It's only after that card has been fully dumped that Metrodroid tries to parse it with agency-specific stuff. There are debug flags in Metrodroid you can use to dump all card communication, but this is left as an exercise for the reader 😄  

As someone who has contributed code to Proxmark: getting this stuff going **on** Proxmark is non-trivial, especially if you have no prior knowledge of the relevant standards – _buying a PC/SC compatible reader is the easiest option._
